### ISSUE TITLE
cf login reports "Error finding available orgs" Server error, status code: 500

### ISSUE THREAD

**Issue by valenbb** [cloudfoundry/cf-release#1043]
CF-Release: 239
Ubuntu stemcell
Openstack CPI

Getting an error from cf login error after successful deploy of cloudfoundry (using signed certs with ha_proxy_z1 job). Here is the full trace:

API endpoint: https://api.cftest.test.local

REQUEST: [2016-08-25T13:54:36Z]
GET /v2/info HTTP/1.1
Host: api.cftest.test.local
Accept: application/json
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.21.0+dff2cf8 / linux

RESPONSE: [2016-08-25T13:54:36Z]
HTTP/1.1 200 OK
Connection: close
Content-Length: 556
Content-Type: application/json;charset=utf-8
Date: Thu, 25 Aug 2016 13:54:33 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: b61c315e-3d22-4912-6c03-c91085982d54
X-Vcap-Request-Id: b61c315e-3d22-4912-6c03-c91085982d54::5b725e39-a7cf-4ee8-8cbb-4f4089a64a14

{"name":"","build":"","support":"http://support.cloudfoundry.com","version":0,"description":"","authorization_endpoint":"https://login.cftest.test.local","token_endpoint":"https://uaa.cftest.test.local","min_cli_version":null,"min_recommended_cli_version":null,"api_version":"2.58.0","app_ssh_endpoint":"ssh.cftest.test.local:2222","app_ssh_host_key_fingerprint":null,"app_ssh_oauth_client":"ssh-proxy","logging_endpoint":"wss://loggregator.cftest.test.local:4443","doppler_logging_endpoint":"wss://doppler.cftest.test.local:4443"}

REQUEST: [2016-08-25T13:54:36Z]
GET /login HTTP/1.1
Host: login.cftest.test.local
Accept: application/json
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.21.0+dff2cf8 / linux

RESPONSE: [2016-08-25T13:54:37Z]
HTTP/1.1 200 OK
Connection: close
Content-Length: 472
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Cache-Control: no-store
Content-Language: en-US
Content-Type: application/json;charset=UTF-8
Date: Thu, 25 Aug 2016 13:54:32 GMT
Expires: 0
Pragma: no-cache
Server: Apache-Coyote/1.1
Strict-Transport-Security: max-age=31536000 ; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Vcap-Request-Id: bacddf27-6324-4612-6762-bfcca2076a3a
X-Xss-Protection: 1; mode=block

{"app":{"version":"3.4.1"},"links":{"uaa":"https://uaa.cftest.test.local","passwd":"https://login.cftest.test.local/forgot_password","login":"https://login.cftest.test.local","register":"https://login.cftest.test.local/create_account"},"zone_name":"uaa","entityID":"login.cftest.test.local","commit_id":"ddae2a1","idpDefinitions":{},"prompts":{"username":["text","Email"],"password":["password","Password"]},"timestamp":"2016-06-20T16:18:23+0000"}
Authenticating...

REQUEST: [2016-08-25T13:54:37Z]
POST /oauth/token HTTP/1.1
Host: login.cftest.test.local
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/x-www-form-urlencoded
User-Agent: go-cli 6.21.0+dff2cf8 / linux

grant_type=password&password=[PRIVATE DATA HIDDEN]&scope=&username=admin

RESPONSE: [2016-08-25T13:54:37Z]
HTTP/1.1 200 OK
Connection: close
Transfer-Encoding: chunked
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Cache-Control: no-store
Content-Type: application/json;charset=UTF-8
Date: Thu, 25 Aug 2016 13:54:32 GMT
Expires: 0
Pragma: no-cache
Pragma: no-cache
Server: Apache-Coyote/1.1
Strict-Transport-Security: max-age=31536000 ; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Vcap-Request-Id: 06f54761-39e3-4a40-4b26-f47cf570f29c
X-Xss-Protection: 1; mode=block

aa1
{"access_token":"[PRIVATE DATA HIDDEN]","token_type":"bearer","refresh_token":"[PRIVATE DATA HIDDEN]","expires_in":599,"scope":"openid scim.read cloud_controller.admin uaa.user cloud_controller.read password.write cloud_controller.write doppler.firehose scim.write","jti":"8a74afc1c0a54e02bf8fb5623a3ab3ed"}
0

OK

REQUEST: [2016-08-25T13:54:37Z]
GET /v2/organizations?order-by=name HTTP/1.1
Host: api.cftest.test.local
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.21.0+dff2cf8 / linux

RESPONSE: [2016-08-25T13:54:37Z]
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 21
Content-Type: text/plain
Date: Thu, 25 Aug 2016 13:54:33 GMT
Server: nginx
X-Vcap-Request-Id: ba576e36-d60c-4604-7568-48e3ed760ede

Internal server error

API endpoint:   https://api.cftest.test.local (API version: 2.58.0)
User:           admin
No org or space targeted, use 'cf target -o ORG -s SPACE'
FAILED
Error finding available orgs
Server error, status code: 500, error code: 0, message: 

Monit summary shows running on all machines.

cf-stub: https://s3.amazonaws.com/andysw/cf-stub.yml

cf-deployment file: https://s3.amazonaws.com/andysw/cf-deployment.yml

Looked over cloud controller logs but could not determine where error is coming from.
In the stub file, I have the org name set.
properties:
  system_domain: cftest.test.local
  system_domain_organization: admin
  app_domains:
   - cftest.test.local

Any suggestion is appreciated.

Note: domain name and certs have been replaced.

**Comment by valenbb** [comment]
Running cf target gives:

API endpoint:   https://api.cftest.test.local (API version: 2.59.0)
User:           admin
No org or space targeted, use 'cf target -o ORG -s SPACE'

**Comment by dsabeti** [comment]
Hi @valenbb. It seems like this is probably an error in the cloud controller. Could you reproduce the error on your environment and then upload the logs? We'll take a look and see if we can find out what's going wrong.

You can download logs from the Cloud Controller by running `bosh logs api_z1`.

**Comment by valenbb** [comment]
Hi @dsabeti, Thanks for taking a look. I did look over the logs but could not determine what went wrong. The URL to where the api_z1 logs have been uploaded (as recent as today): 
https://s3.amazonaws.com/andysw/api_z1.0.2016-08-31-11-36-27.tar.gz

**Comment by jaresty** [comment]
Hi @valenbb,

Can you please try running `CF_TRACE=true cf curl /v2/info` and `CF_TRACE=true cf running-environment-variable-group` (because the latter requires authorized access and we want to check your authorization flow). From looking at the logs, it appears the request is making it as far as nginx on the Cloud Controller, but we did not see the request be logged by the Cloud Controller job, so it may be that something is happening between nginx and the CC.

Thanks,
Josh &&  @APShirley

**Comment by valenbb** [comment]
Hi Josh && @APShirley,

Here is the requested info:

root@cfboshinit:~# CF_TRACE=true cf curl /v2/info

REQUEST: [2016-09-02T02:03:01Z]
GET /v2/info HTTP/1.1
Host: api.cftest.test.local
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.21.1+6fd3c9f / linux



RESPONSE: [2016-09-02T02:03:01Z]
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 21
Content-Type: text/plain
Date: Fri, 02 Sep 2016 02:02:55 GMT
Server: nginx
X-Vcap-Request-Id: 754fca55-4c51-4a44-68cf-5a878e08a0e4

Internal server error

root@cfboshinit:~# CF_TRACE=true cf running-environment-variable-group
Retrieving the contents of the running environment variable group as alalla...

REQUEST: [2016-09-02T02:05:02Z]
GET /v2/config/environment_variable_groups/running HTTP/1.1
Host: api.cftest.test.local
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.21.1+6fd3c9f / linux



RESPONSE: [2016-09-02T02:05:02Z]
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 21
Content-Type: text/plain
Date: Fri, 02 Sep 2016 02:04:56 GMT
Server: nginx
X-Vcap-Request-Id: 56a35ce9-cec8-4028-6e20-876bb517c7ca

Internal server error
FAILED
Server error, status code: 500, error code: 0, message:

**Comment by eflynn00** [comment]
As a followup to @valenbb cf trace above, this is the output from nginx_cc's nginx.access.log when running the commands;

**CF_TRACE=true cf running-environment-variable-group**

api.cftest.somesubdomain.somedomain.com - [02/Sep/2016:15:10:38 +0000] "GET /v2/config/environment_variable_groups/running HTTP/1.1" 500 195 "-" "go-cli 6.21.0+dff2cf8 / darwin" 3.39.127.229, 172.16.34.125, 172.16.34.130 vcap_request_id:- response_time:0.004

**CF_TRACE=true cf curl /v2/info** 

api.cftest.somesubdomain.somedomain.com - [02/Sep/2016:15:13:27 +0000] "GET /v2/info HTTP/1.1" 500 195 "-" "go-cli 6.21.0+dff2cf8 / darwin" 3.39.127.229, 172.16.34.125, 172.16.34.130 vcap_request_id:- response_time:0.003

**Comment by jaresty** [comment]
@eflynn00 @valenbb Can you try running `cf logout` and then run the commands again? I'm suspicious that the exception could be happening because you are logged in.  `CF_TRACE=true cf curl /v2/info` and `CF_TRACE=true cf running-environment-variable-group`

**Comment by eflynn00** [comment]
@jaresty, I logged out, and ran CF_TRACE=true cf curl /v2/info. It was successful. 

Here is the cf trace output;

> someuser$ CF_TRACE=true cf curl /v2/info

REQUEST: [2016-09-02T13:02:17-04:00]
GET /v2/info HTTP/1.1
Host: api.somesubdomain.somedomain.com
Accept: application/json
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.21.0+dff2cf8 / darwin



RESPONSE: [2016-09-02T13:02:17-04:00]
HTTP/1.1 200 OK
Connection: close
Content-Length: 556
Content-Type: application/json;charset=utf-8
Date: Fri, 02 Sep 2016 17:02:17 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: d2804667-afa2-4860-5723-75a10ba59410
X-Vcap-Request-Id: d2804667-afa2-4860-5723-75a10ba59410::9faaf1c8-3792-463c-b4e0-716d0e23874f

{"name":"","build":"","support":"http://support.cloudfoundry.com","version":0,"description":"","authorization_endpoint":"https://login.somesubdomain.somedomain.com","token_endpoint":"https://uaa.somesubdomain.somedomain.com","min_cli_version":null,"min_recommended_cli_version":null,"api_version":"2.59.0","app_ssh_endpoint":"ssh.somesubdomain.somedomain.com:2222","app_ssh_host_key_fingerprint":null,"app_ssh_oauth_client":"ssh-proxy","logging_endpoint":"wss://loggregator.somesubdomain.somedomain.com:4443","doppler_logging_endpoint":"wss://doppler.somesubdomain.somedomain.com:4443"}

And, here's the corresponding nginx_cc log output;

> api.somesubdomain.subdomain.ge.com - [02/Sep/2016:17:02:17 +0000] "GET /v2/info HTTP/1.1" 200 841 "-" "go-cli 6.21.0+dff2cf8 / darwin" 10.1..127.229, 172.16.34.125, 172.16.34.130 vcap_request_id:d2804667-afa2-4860-5723-75a10ba59410::9faaf1c8-3792-463c-b4e0-716d0e23874f response_time:0.007

Although, when logged out, and running, CF_TRACE=true cf running-environment-variable-group, it fails. But, I think this is expected, since environment variable group needs authentication and authorization. Here's the console output, since there are no logs in nginx_cc;

> someuser$ CF_TRACE=true cf running-environment-variable-group
FAILED
Not logged in. Use 'cf login' to log in.

**Comment by valenbb** [comment]
We have resolved this issue. Error was due to jwt verification key. The documentation for editing the cf-stub.yml - jwt section does not tell us what to do properly. We could not use the generated pub file.
Used the following the cat the output file:
openssl rsa -in jwt-key.pem -pubout > key.pub

ssh-keygen -f jwt-key.pem does create a pub key that includes the begin and end lines.

jwt:
      verification_key: JWT_VERIFICATION_KEY
      signing_key: JWT_SIGNING_KEY

Instead of just stating where each key is used, the begin and end lines should be inserted so that admins know what to look for after generating these keys. For example,

verification_key: JWT_VERIFICATION_KEY
         -----BEGIN PUBLIC KEY-----
         PUBLIC_KEY
          -----END PUBLIC KEY-----
signing_key: JWT_SIGNING_KEY
          -----BEGIN RSA PRIVATE KEY-----
          RSA_PRIVATE_KEY
          -----END RSA PRIVATE KEY-----
